Kernel/FileSystem/ISO9660FileSystem.cpp

```diff
@@ -0,0 +1,695 @@
+            // records, then the leftover bytes in the logical block are
+            // all zeros. So if our directory header has a length of 0,
+            // we're probably looking at padding.
+            //
+            // Of course, this doesn't mean we're done; it only means that there
+            // are no more directory entries in *this* logical block.  If we
+            // have at least one more logical block of data length to go, we
+            // need to snap to the next logical block, because directory records
+            // cannot span multiple logical blocks.
+            u32 remaining_bytes = m_current_directory.entry->length - m_current_directory.offset;
+            if (remaining_bytes > m_fs.logical_block_size()) {
+                m_current_directory.offset += remaining_bytes % m_fs.logical_block_size();
+                get_header();
+
+                dbgln_if(ISO9660_VERY_DEBUG, "skip(): Snapped to next logical block (succeeded)");
+                return true;
+            }
+
+            dbgln_if(ISO9660_VERY_DEBUG, "skip(): Was at the last logical block, at padding now (offset {}, data length {})", m_current_directory.entry->length, m_current_directory.offset);
+            return false;
+        }
+
+        dbgln_if(ISO9660_VERY_DEBUG, "skip(): Skipped to next item");
+        return true;
+    }
+
+    bool go_up()
+    {
+        if (m_directory_stack.is_empty()) {
+            dbgln_if(ISO9660_VERY_DEBUG, "go_up(): Empty directory stack");
+            return false;
+        }
+
+        m_current_directory = m_directory_stack.take_last();
+        get_header();
+
+        dbgln_if(ISO9660_VERY_DEBUG, "go_up(): Went up a directory");
+        return true;
+    }
+
+    bool done()
+    {
+        VERIFY(m_current_directory.entry);
+        auto result = m_directory_stack.is_empty() && m_current_directory.offset >= m_current_directory.entry->length;
+        dbgln_if(ISO9660_VERY_DEBUG, "done(): {}", result);
+        return result;
+    }
+
+private:
+    KResult read_directory_contents()
+    {
+        auto result = m_fs.directory_entry_for_record({}, m_current_header);
+        if (result.is_error()) {
+            return result.error();
+        }
+
+        m_current_directory.entry = result.value();
+        return KSuccess;
+    }
+
+    void get_header()
+    {
+        VERIFY(m_current_directory.entry);
+        if (!m_current_directory.entry->blocks)
+            return;
+
+        m_current_header = reinterpret_cast<ISO::DirectoryRecordHeader const*>(m_current_directory.entry->blocks->data() + m_current_directory.offset);
+    }
+
+    ISO9660FS& m_fs;
+
+    DirectoryState m_current_directory;
+    ISO::DirectoryRecordHeader const* m_current_header { nullptr };
+
+    Vector<DirectoryState> m_directory_stack;
+};
+
+NonnullRefPtr<ISO9660FS> ISO9660FS::create(FileDescription& description)
+{
+    return adopt_ref(*new ISO9660FS(description));
```

Comment:
This can't return a `NonnullRefPtr<...>` someone can mount an ISO at anytime.. hence, we need this to be OOM safe.

Reply:
Check `Ext2FileSystem.cpp`, it will make you sadder. :^) I'm fixing my instance though.